### ISSUE TITLE
Is there a use case for customizing default controls?

### ISSUE THREAD

**Issue by avayvod** [WICG/controls-list#5]
We believe customizing default controls is useful for both users and websites (compared to implementing custom controls) for the following reasons:
- custom controls exist in the wild, thus proving there's need for customized controls;
- the UA is in full control of the media experience thus making the default controls potentially better than any custom controls website can implement;
- most websites implementing custom controls will do a worse job than the UA, esp. in terms of cross-browser support or accessibility support.

A more detailed ongoing discussion is happening on the HTML spec [issue](https://github.com/whatwg/html/issues/2293).

**Comment by zendive** [comment]
Hello @avayvod, I work for the online video creation platform and we heavily dependent on native video player in general and Chrome in particular.

We use video element for **editorial** purposes both as a simple video preview at various dimensions: full-page, 688x387px, 284x160px, and as background, programmatically managed, media item in custom video player.

In our platform I see several use cases for customizing default controls that may bring positive effect:
1. removing download control button from players where there is no intent to enable download because of content restrictions from media providers.
2. removing download control button because video player has low-resolution source (for the web) and a nearby dedicated button links to high resolution video source.
3. removing any other controls except for play button and seek-bar on small players like 284x160px will widen the seek-bar to the full length allowing more space for the interaction with it [seek-bar]. Currently it is possible to enter full screen in order to make a fine-seek tuning but having in mind that we have other controls on the UI that should be visible (not ads) and low-resolution video becomes pixelated - makes it undesired.

> most websites implementing custom controls will do a worse job than the UA, esp. in terms of cross-browser support or accessibility support.

Cross-browser fullscreen support is the trickiest part in that I think.

Thank you

**Comment by johnBartos** [comment]
>- custom controls exist in the wild, thus proving there's need for customized controls;

There is a need for customized controls, but I don't think that a browser can or should cover all of use cases which would make an OVP adopt a controls API. Is this spec prepared to provide controls for Ads? In addition, bespoke customized controls to give us the flexibility to add and remove things which aren't necessarily a control (i.e. "next up" modal, rewind button) but which behave like one. Furthermore, they give our publishers the ability to customize their controls and brand them to meet the desires of their userbase or product goals. 

>- the UA is in full control of the media experience thus making the default controls potentially better than any custom controls website can implement;

Better for who? I agree that some players do not make very good controls and for those users forcing to native would be an improvement. However, there are several controls which are not default (like a rewind Nseconds button) which users would be missing if they do not use custom controls. Giving the users an option (which will probably be poorly understood by the casual user) to override a well-done controls layer would be a downgrade in UX.

>- most websites implementing custom controls will do a worse job than the UA, esp. in terms of cross-browser support or accessibility support.

Some do and some don't, but whose to say Google is the best at implementing controls which meets the requirements of every OVP and publisher? And who is the arbiter of "good UX" and "bad UX", when UX paradigms change frequently? I don't think Google should be in the business of deciding what is best for us. If your controls suck, then people won't use your player.


To summarize, I think this proposal falls short because:
1. This proposal does not and should not provide contingencies for every control players implement (ads, rewind, etc)
2. Giving users a confusing choice between default & custom hurts users in the case where custom controls are well-implemented and offer additional features
3. UX is the responsibility of the player and publisher to implement correctly. I don't think Google should be the arbiter of what looks good and what doesn't.

**Comment by avayvod** [comment]
@johnBartos 

> I don't think that a browser can or should cover all of use cases which would make an OVP adopt a controls API

I agree. If the website has a use case not supported by the default controls, they should use custom controls as a valid option. We claim there's a lot of cases where websites don't need additional controls but do want to disable some control that the UA would provide by default.

Also for fast forward/rewind (also previous and next track), check out the [MediaSession API](https://wicg.github.io/mediasession/). It doesn't affect the native controls though.

> Giving the users an option (which will probably be poorly understood by the casual user) to override a well-done controls layer would be a downgrade in UX.

Well, unless the context menu is disabled, there's an option to enable default controls for the case when the custom controls are missing or not working. However I don't imagine users going that path if the custom controls are done well.

This API is not about giving this option. If the website uses custom controls, it uses custom controls.
If it doesn't have resources to implement the full solution but needs customization, it uses the API.

The choice is for the web developers, not the users.

> Better for who? ... I don't think Google should be in the business of deciding what is best for us.

The intended use for the API is for the websites that are okay with the overall default controls UX modulo simple modifications (like removing a button).

**Comment by johnBartos** [comment]
@avayvod Fair points, I was speaking from the perspective of a player developer and from what I gather I'm not the intended audience 😅. Thinking as a smaller/independent publisher I think that this is a pretty good idea. 

My only concern is that this API is kind of an all-or-nothing deal: it's either the case that you're able to customize exactly what you want (in which case this API is perfect for you), otherwise you need to implement custom controls. And it's not just features which people consider - it's support too. IE is a non-negligible portion of video watchers and they would never get this API. If the publisher decides IE support is important then it would be hard to sell an API that only works in one browser.

I think I'm still misunderstanding this point though:
>This API is not about giving this option. If the website uses custom controls, it uses custom controls.

So users don't have the ability to choose default/custom controls, it's the publishers who do?

**Comment by avayvod** [comment]
@johnBartos 

> So users don't have the ability to choose default/custom controls, it's the publishers who do?

Yes.

**Comment by jimmywarting** [comment]
> * the UA is in full control of the media experience thus making the default controls potentially better than any custom controls website can implement;
> * most websites implementing custom controls will do a worse job than the UA, esp. in terms of cross-browser support or accessibility support.

I fully agree with you. I like the native controllers and other dose a worse job. But if I have HLS or dash stream, then I need to go full custom so that i can add something like a quality picker for 720p, 1080p or auto. Having to reimplement everything sucks if would like to add something else into the mix.
